TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+}
+```
+
+This would remove the intricated structure we have with builders and its children.
+
+A side effect of this approach, because we are dealing with functions (versus structs) we are able to remove local state. An example of such case is `enableClinicalDownloads` at `ClinicalRecordsChildBuilders`. You can see the code smell here:
+
+```swift
+builders.appointmentList(with: session,
+                         enableClinicalDownloads: builders.enableClinicalDownloads,
+                         addressSwitches: builders.addressSwitches)
+```
+
+With the free functions approach, this would never be possible to do, since builders don't hold state:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeAppointmentList(session: UserSession,
+                                 enableClinicalDownloads: Bool,
+                                 addressSwitches: AddressSwitchesProtocol,
+                                 primary: Flow,
+                                 secondary: Flow) -> UIViewController
+    }
+}
+```
+
+At the call site, we would replace this:
+
+```swift
+case let .appointmentList(session):
+    builders
+            .appointmentList(with: session,
+                             enableClinicalDownloads: builders.enableClinicalDownloads,
+                             addressSwitches: builders.addressSwitches)
+            .make(primary: navigation, secondary: modal)
+            |> navigation.present
+```
+
+With this:
+
+```swift
+case let .appointmentList(session):
+    Builder.ClinicalRecords
+            .makeAppointmentList(with: session,
+                                enableClinicalDownloads: enableClinicalDownloads,
+                                addressSwitches: addressSwitches,
+                                primary: navigation,
+                                secondary: modal)
+            |> navigation.present
+```
+
+## Impact on an existing codebase
+
+As Ilya pointed out it would make unit testing a bit more cumbersome. That's the case because you no longer inject builders. I would like to make the case for the following:
+
+1. We don't test `FlowControllers` with Unit tests anyway. This is not a reason to not do it, it's simply difficult to justify time doing so, when there are more important pieces to be tested and those tests would bring, in my opinion, little value. More value would be achieved by the next point (**2.**).
+2. We can test that a particular navigation/flow is done as expected independently of this proposal being approved or not. For that to happen we simply observe a ViewModel state and assert on it, since we derive routes from the state.
+3. If we "really" want to test FlowControllers, it's still possible by mocking `Current`, which is something trivial to achieve (`World.init()`). The difference here is that we can't say we are **Unit** testing FlowControllers, but instead these become **Integration** tests. I personally don't feel this is semantically significant. 
```

Comment:
By mocking `Current` we will be only able to mock the network session (assuming it is there which is not the case ATM). Builder methods provide view controllers as a black box and do not provide any access to underlying view models so there is no way to act on them in tests if builder is not injected or can't be overridden in any other way. So we wouldn't be able for example to write a test that would trigger a UI action in presented view model that is supposed to dismiss its screen and then assert that flow controller actually dismisses it by calling `dismiss` on a flow.

We currently rely heavily on automation tests for testing flows, but it does not mean IMO that we should not look into unit testing flow controllers. Unit tests have important advantages over automation tests (easier to write, faster to execute), which does not mean we should drop automation in favour of unit tests either. We should use any tools at our disposal (in a suitable way ofc) to ensure quality.

Reply:
Based on Anders suggestion, we would mock at a higher level, not at the network level. As I previously said, one thing is to have an aspiration to do something which is what your critique aims at:

> So we wouldn't be able for example to write a test that would trigger a UI action in presented view model that is supposed to dismiss its screen and then assert that flow controller actually dismisses it by calling dismiss on a flow.

You can still write your tests, integrations tests in this case, and they would be as fast as any other unit test since they don't rely on UI as well. I am not going to spend more time debating this point, simply because I already provided an alternative for testing, that doesn't rely on UI Automation.

Reply:
Then please provide an example of such test.

Reply:
We are not doing any sort of those tests, so there isn't even an accepted way across the team on how to test, which is precisely my point. In any case:


```swift
world = World()
world.userSession = MockedUserSession() 

var isCalled = false
let mockedFlow = { isCalled = true }
let shouldntBeCalled = { fatalError() }

let flowController = AFlowController(navigation: mockedFlow, model: shouldntBeCalled)
flowController.handle(.aState)

XCTAssert(isCalled == true)
```

This would test that the right flow is presented.

Reply:
My concern is not about a `handle` method that is being tested in this example, but about testing how flow controller reacts on UI actions from presented view models.

Reply:
It feels that you are talking about integration tests, where you assert an entire flow from a ViewModel action down to a flowController "reaction". Assuming @sergdort 's proposal would be accepted (again this is a weak argument from my side, because it's a big assumption), you could just inject mocked FlowControllers Feedback into the VM and assert on those.

Reply:
I'm talking about unit testing flow controller in isolation of their environment like real builders and other real view models. Injecting mocked feedback is something one would do in unit tests of view model, not flow controller which would define such feedback. This feedback is something that should be tested (as we test other feedbacks, why wouldn't we test some of them)

Reply:
Not sure what you want to achieve with testing FlowControllers in isolation. to make sure you get the right method being called in a builder? Is that it?

Reply:
That can be tested by passing the dismiss state to the `handler` and asserting on the flow.

Reply:
This would be a state of a presented view model, where and how it will be passed to the flow controller that presented it?  There will be no access to this view model from the test.

Reply:
You can pass via a handler as part of the state, that would be called once the presented screen is dismissed. For example `.showScreen(someData, dismissHandler)`

Reply:
How this would help to test that `flow.dismiss` is called in the following implementation?

```swift
func handle(_ route: Route) {
  switch route {
    case let .showScreen(someData, dismissHandler):
       Builder.Something.makeSomething(someData, onDismiss: { result in
          flow.dismiss() { dismissHandler(result) }
       } |> flow.present
  }
}
```

Reply:
You assert that `dismissHandler` is called. 🤷‍♂️

Reply:
the fact that `dismissHandler` has little to do with ensuring that `dismiss` (or any other method, it can be `popToRoot` or something) is called on the proper flow (flow most likely will be a mock implementation as well). Also one could write it as

```
flow.dismiss()
dismissHandler()
```

This may be completely fine implementation in some cases, but not all. So here `dismissHandler` will be caller regardless of `flow.dismiss` being called at all or being called on a correct flow instance.

Reply:
From `FC` perspective you usually what to unit test that providing a `route` it calls correct method on builder and correct `flow`. I not sure what else we would be interested in?

Reply:
The builder method, with the current proposal, wouldn't be possible in a straightforward way. The flow calling would be.

Reply:
@ilyapuchka quoting yourself:

> I'd be fine dismissing my concern about testability

I think we already agreed that:

1. We are not testing those particular scenarios, currently, independently of this proposal
2. We have no plans to do so in the short/medium term.
3. Critical paths of the App are already covered with UI Automation.

I am unsure why we keep going around this point.

Reply:
The example shared shows how to test FC if/when needed, the SUT needs to be atomic so we either test the FC or the VM but not both at the same time, that enters the territory of Acceptance/Integration Tests where UI Tests have clear advantages as you don't need to be aware of all the underlying implementation.